solr/solrj/src/java/org/apache/solr/client/solrj/impl/CloudHttp2SolrClient.java

```diff
@@ -451,12 +478,19 @@ public Builder withZkClientTimeout(int zkClientTimeout, TimeUnit unit) {
 
     /** Create a {@link CloudHttp2SolrClient} based on the provided configuration. */
     public CloudHttp2SolrClient build() {
-      if (!zkHosts.isEmpty() && !solrUrls.isEmpty()) {
+      int providedOptions = 0;
+      if (!zkHosts.isEmpty()) providedOptions++;
+      if (!solrUrls.isEmpty()) providedOptions++;
+      if (stateProvider != null) providedOptions++;
+
+      if (providedOptions > 1) {
         throw new IllegalArgumentException(
-            "Both zkHost(s) & solrUrl(s) have been specified. Only specify one.");
-      } else if (zkHosts.isEmpty() && solrUrls.isEmpty()) {
-        throw new IllegalArgumentException("Both zkHosts and solrUrl cannot be null.");
+            "Only one of zkHost(s), solrUrl(s), or stateProvider should be specified.");
+      } else if (providedOptions == 0) {
+        throw new IllegalArgumentException(
+            "One of zkHosts, solrUrls, or stateProvider must be specified.");
       }
```

Comment:
As you are separating out a Factory for the ClusterStateProvider, I wonder if this code here might move over there as it's tightly correlated with it?  Not important; just a thought.

Reply:
Yes, actually it can be moved there. Totally make sense there. I guess we can move the whole factory logic along with this in builder.